arch/arc/include/asm/atomic.h

```diff
@@ -69,41 +71,81 @@ static inline int atomic_fetch_##op##_relaxed(int i, atomic_t *v)		\
 	"	scond   %[val], [%[ctr]]		\n"		\
 	"	bnz     1b				\n"		\
 	: [val]	"=&r"	(val),						\
-	  [orig] "=&r" (orig)						\
+	  [orig] "=&r" (orig),						\
+		"+m"	(v->counter)					\
 	: [ctr]	"r"	(&v->counter),					\
 	  [i]	"ir"	(i)						\
 	: "cc");							\
 									\
 	return orig;							\
 }
 
-#define atomic_fetch_add_relaxed	atomic_fetch_add_relaxed
-#define atomic_fetch_sub_relaxed	atomic_fetch_sub_relaxed
+#ifdef CONFIG_ARC_HAS_ATLD
+#define ATOMIC_FETCH_ATLD_OP(op, asm_op)				\
+static inline int atomic_fetch_atld_##op##_relaxed(int i, atomic_t *v)	\
+{									\
+	unsigned int orig = i;						\
+									\
+	__asm__ __volatile__(						\
+	"	atld."#asm_op" %[orig], [%[ctr]]	\n"		\
+	: [orig] "+r"(orig), "+m"(v->counter)				\
+	: [ctr] "r"(&v->counter)					\
```

Comment:
I don't think you need this, you already used it in the memory constraint.

Reply:
Done